### PR TITLE
pkp/pkp-lib#2072 Prototype of versioning based on new publication entity

### PR DESCRIPTION
**Pull request by NateWr**
This commit includes all of the initial work done to support versioning based on a split between submissions and publications. All submission data related to publication, such as title, abstract, citations, authors and galleys, has been moved to a new publication entity.

Submissions have a "one to many" relationship to Publications. Each Submission may have one or more Publications attached to it. Each Publication is treated as a new version. Published version data can not be modified.

- New publication entity split from submissions
- New API endpoints for publications
- Workflow UI changes to support versions (publications)
- Pre-publication validation checks
- New STATUS_SCHEDULED for publications scheduled for publication in a future issue
- Deprecated many methods on the Submission object
- Upgrade scripts written from 3.1.x.
- Tests updated to work, except for issue import.

Some code is commented out or has not been updated yet. Progress on remaining support for versioning will be tracked in Github.

See: https://github.com/pkp/pkp-lib/projects/15

### CODE REVIEW

**Comment by asmecher** [comment]
:fireworks:!

File: classes/plugins/DOIPubIdExportPlugin.inc.php [hunk 1]
```diff
@@ -127,18 +127,18 @@ function getAdditionalFieldNames($hookName, $args) {
...
-			$this->getPubIdType(),
-			null,
-			null,
-			null,
-			$this->getPluginSettingsPrefix(). '::' . DOI_EXPORT_REGISTERED_DOI,
-			null,
-			null
-		);
-		return $articles->toArray();
+		// $publishedSubmissionDao = DAORegistry::getDAO('PublishedSubmissionDAO'); /* @var $publishedSubmissionDao PublishedSubmissionDAO */
```
Comment (NateWr) [inline classes/plugins/DOIPubIdExportPlugin.inc.php]: To be completed as part of this issue: https://github.com/pkp/pkp-lib/issues/4867

File: classes/plugins/PubObjectsExportPlugin.inc.php [hunk 2]
```diff
@@ -432,19 +432,19 @@ function callbackParseCronTab($hookName, $args) {
...
-			null,
-			null,
-			null,
-			null,
-			$this->getDepositStatusSettingName(),
-			EXPORT_STATUS_NOT_DEPOSITED,
-			null
-		);
-		return $articles->toArray();
+		// // Retrieve all published submissions that have not yet been registered.
```
Comment (NateWr) [inline classes/plugins/PubObjectsExportPlugin.inc.php]: Filed for completion here: https://github.com/pkp/pkp-lib/issues/4867

File: classes/services/GalleyService.inc.php [hunk 3]
```diff
@@ -15,172 +15,197 @@
...
-							$parentPath,
-							'view',
-							array($parentId, $galley->getBestGalleyId())
-						);
-					}
-					break;
-				case 'seq':
-				case 'sequence':
-					$values[$prop] = $galley->getSequence();
+					// TODO
```
Comment (NateWr) [inline classes/services/GalleyService.inc.php]: Filed for fixing here: https://github.com/pkp/pkp-lib/issues/4886

File: controllers/grid/pubIds/PubIdExportSubmissionsListGridHandler.inc.php [hunk 4]
```diff
@@ -34,16 +33,17 @@ protected function loadData($request, $filter) {
...
-			$context->getId(),
-			$this->_plugin->getPubIdType(),
-			$title,
-			$author,
-			$issueId,
-			$pubIdStatusSettingName,
-			$statusId,
-			$this->getGridRangeInfo($request, $this->getId())
-		);
+		return [];
```
Comment (NateWr) [inline controllers/grid/pubIds/PubIdExportSubmissionsListGridHandler.inc.php]: Filed for completion here: https://github.com/pkp/pkp-lib/issues/4867

File: pages/article/ArticleHandler.inc.php [hunk 5]
```diff
@@ -108,19 +131,24 @@ function view($args, $request) {
 		));
 		$this->setupTemplate($request);
 
-		if (!$this->userCanViewGalley($request, $articleId, $galleyId)) fatalError('Cannot view galley.');
+		// TODO: this defaults to the current publication but should
+		// retrieve the publication requested from the URL if it is
+		// passed as an arg
```
Comment (NateWr) [inline pages/article/ArticleHandler.inc.php]: Filed for completion here: https://github.com/pkp/pkp-lib/issues/4870

File: pages/article/ArticleHandler.inc.php [hunk 6]
```diff
@@ -143,40 +171,34 @@ function view($args, $request) {
...
+			'section' => DAORegistry::getDAO('SectionDAO')->getById($requestedPublication->getData('sectionId')),
+			'licenseTerms' => $context->getLocalizedData('licenseTerms'),
+			'licenseUrl' => $requestedPublication->getData('licenseUrl'),
+			'ccLicenseBadge' => Application::get()->getCCLicenseBadge($requestedPublication->getData('licenseUrl')),
+			'copyrightHolder' => $requestedPublication->getData('copyrightHolder'),
+			'copyrightYear' => $requestedPublication->getData('copyrightYear'),
+			'pubIdPlugins' => PluginRegistry::loadCategory('pubIds', true),
+			'parsedCitations' => $parsedCitations ?? [],
+			// @TODO
+			// 'keywords' => ...,
```
Comment (NateWr) [inline pages/article/ArticleHandler.inc.php]: Filed for completion: https://github.com/pkp/pkp-lib/issues/5030

File: tests/data/60-content/ImportIssueTest.php [hunk 7]
```diff
@@ -37,7 +37,13 @@ function testImportIssue() {
...
 
 		// Ensure that the import was listed as completed.
-		$this->waitForElementPresent('//*[contains(text(),\'The import completed successfully.\')]//li[contains(text(),\'Vol 1 No 1\')]');
+		// TODO: Import/export is not yet compatible with versioning.
+		// See: https://github.com/pkp/pkp-lib/issues/4880
+		//
+		// Because of this problem, the publish issue tests (jmwandenga/vkarbasizaed) were
+		// updated to put the articles in Vol 1 No 2, instead of Vol 1 No 1. This may
+		// need to be corrected after import/export is fixed.
+		// $this->waitForElementPresent('//*[contains(text(),\'The import completed successfully.\')]//li[contains(text(),\'Vol 1 No 1\')]');
```
Comment (NateWr) [inline tests/data/60-content/ImportIssueTest.php]: Filed for fixing: https://github.com/pkp/pkp-lib/issues/4880

**Pull request merged**